include/misc/mempool.h

```diff
@@ -0,0 +1,59 @@
+/*
+ * Copyright (c) 2018 Intel Corporation
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+#ifndef SYS_MEMPOOL_H
+#define SYS_MEMPOOL_H
+
+#include <kernel.h>
+#include <misc/mempool_base.h>
+
+struct sys_mem_pool {
+	struct sys_mem_pool_base base;
+	struct k_mutex *mutex;
+};
+
+/*
+ * @brief Create a memory pool from a chunk of RAM
+ *
+ * Given an arbitrary blob of memory, carve this memory up such that it
+ * contains all the control data structures associated with a memory pool
+ * and all remaining storage is used for the pool itself.
+ *
+ * @param buf Memory region to use for the pool
+ * @param size Size of the provided memory region
+ * @param minsz Size of the smallest blocks in the pool
+ * @param maxsz Size of the largest blocks in the pool
+ * @param mutex Initialized k_mutex object that the pool will use for
+ *              concurrency control.
+ * @return An initialized sys_mem_pool object pointer
+ */
+struct sys_mem_pool *sys_mem_pool_create(char *buf, size_t size,
+					 size_t minsz, size_t maxsz,
+					 struct k_mutex *mutex);
+
```

Comment:
Random nit: if we're writing a new API, we might consider calling this a "heap" instead of a "memory pool".  The industry already has a well-understood name for this kind of thing, and it saves like seven bytes of symbol name typing.